[kennethsoh] iP

kennethsoh opened the pull request:
# Duck 🐥

> "Quack" – every duck that ever existed

Duck helps you _organise_ and _streamline_ your tasks effectively. It is:

- User-friendly
- _Efficient_
- ~Slow~ **Super fast** 

All you need to do is:

1. Download it from [here](https://github.com/nus-cs2103-AY2425S2/ip/pull/83).
2. Install it on your computer.
3. Add your tasks.
4. Let it do the rest for you! 

And the best part? It's **FREE**!

### Features:
- [ ] Task management
- [ ] Deadline tracking (coming soon)
- [ ] Reminders (coming soon)

If you're a developer, you can use this to practice your programming skills. Here's an example of the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kennethsoh committed:
Level 0

kennethsoh committed:
level 1

kennethsoh committed:
level 2

kennethsoh committed:
Level-3, A-Classes

kennethsoh committed:
Level-4, A-Inheritance

kennethsoh committed:
A-TextUiTesting

kennethsoh committed:
Level-5, A-Exceptions

kennethsoh committed:
Level-6

kennethsoh committed:
Complete Level-7

kennethsoh committed:
Merge branch 'branch-Level-7'

kennethsoh committed:
Complete Level-8

kennethsoh committed:
Merge branch 'branch-Level-8'

kennethsoh committed:
Merge branch 'add-gradle-support'

kennethsoh committed:
Complete A-Packages, A-Gradle

kennethsoh committed:
Complete A-MoreOOP

kennethsoh committed:
Merge branch 'branch-A-Increments'

kennethsoh committed:
Complete A-JUnit

kennethsoh committed:
Complete A-Jar

kennethsoh committed:
Complete Level-9

kennethsoh committed:
Complete A-CodingStandard

kennethsoh committed:
Complete A-JavaDoc

kennethsoh committed:
Merge branch 'branch-A-JavaDoc'

kennethsoh committed:
Merge branch 'branch-A-CodingStandard'

kennethsoh committed:
Merge branch 'branch-Level-9'

kennethsoh committed:
Fix Style and JavaDoc

kennethsoh committed:
Complete A-CheckStyle

kennethsoh committed:
Merge branch 'branch-A-CheckStyle'

kennethsoh committed:
Add LocalDateTime.now() feature to deadline and event tasks

kennethsoh committed:
Complete Level-10

kennethsoh committed:
Merge branch 'branch-Level-10'

kennethsoh committed:
Fix JavaDoc and add 'later' time feature

kennethsoh committed:
Allow automatic element resizing

caando commented on src/main/java/duck/Storage.java:
```diff
@@ -0,0 +1,107 @@
+            saveFile.getParentFile().mkdirs();
+            return list;
+        } else {
+            BufferedReader reader = new BufferedReader(new FileReader(this.filePath));
```
nit: Should `BufferedReader` be closed after use?

caando commented on src/main/java/duck/Storage.java:
```diff
@@ -0,0 +1,107 @@
+     * @throws IOException If an error occurs while reading the file.
+     */
+    public ArrayList<Task> load() throws IOException {
+        ArrayList<Task> list = new ArrayList<Task>();
```
nit: Should diamond operator be used?

```suggestion
        ArrayList<Task> list = new ArrayList<>();
```

caando commented on src/main/java/duck/Storage.java:
```diff
@@ -0,0 +1,107 @@
+                boolean isDone = parts[1].trim().equals("1");
+                String description = parts[2].trim();
+
+                if (type.equals("T")) {
+                    list.add(new ToDo(isDone, description));
+                } else if (type.equals("D")) {
+                    String deadline = parts[3].trim();
+                    LocalDateTime by = LocalDateTime.parse(deadline, dateTimeFormat);
+                    list.add(new Deadline(isDone, description, by.format(dateTimeFormat)));
+                } else if (type.equals("E")) {
+                    String[] details = parts[3].trim().split(" to ");
+                    String from = details[0].trim();
+                    String to = details[0].trim();
+                    LocalDateTime fromTime = LocalDateTime.parse(from, dateTimeFormat);
+                    LocalDateTime toTime = LocalDateTime.parse(to, dateTimeFormat);
+                    list.add(new Event(isDone, description,
+                            fromTime.format(dateTimeFormat), toTime.format(dateTimeFormat)));
+                }
```
nit: Should this be converted to a switch statement?

caando commented on src/main/java/duck/Ui.java:
```diff
@@ -0,0 +1,128 @@
+     */
+    public String mark(Task t) {
+        return "Nice! I've marked this task as done:\n"
+               + "     [" + t.getStatusIcon() + "] " + t.getDescription() + "\n";
```
nit: Should line wrapping start with 8 spaces? I think alignment with previous line is fine too.

```suggestion
                + "     [" + t.getStatusIcon() + "] " + t.getDescription() + "\n";
```

caando commented on src/main/java/duck/Parser.java:
```diff
@@ -0,0 +1,252 @@
+        } catch (IndexOutOfBoundsException e) {
+            return ui.showErrorMessage(e.getMessage());
+        } catch (Exception e) {
+            return ui.showErrorMessage("Invalid Input, Try again!");
```
Should specifically `RuntimeException` be caught here?
```suggestion
        } catch (RuntimeException e) {
```

caando commented on src/main/java/duck/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return The formatted deadline task string.
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + by.format(DateTimeFormatter.ofPattern("dd MMM yyyy, hh:mma")) + ")";
+    }
```
Should line length be kept under 120 characters?

nit: Any reason why you prefer string concatenation over `String.format`? 

```suggestion
    @Override
    public String toString() {
        return String.format("[D]%s (by: %s)", super.toString(), 
                by.format(DateTimeFormatter.ofPattern("dd MMM yyyy, hh:mma")));
    }
```

caando commented on src/main/java/duck/Event.java:
```diff
@@ -0,0 +1,62 @@
+package duck;
+import java.time.LocalDateTime;
```
Should there be a new line separator between package and imports?

```suggestion
package duck;

import java.time.LocalDateTime;
```

caando commented on src/main/java/duck/ToDo.java:
```diff
@@ -0,0 +1,29 @@
+package duck;
+
+/**
+ * Represents a simple ToDo task.
+ */
+
+public class ToDo extends Task {
```
Should the start of class immediately follow the end of javadoc? 

```suggestion
/**
 * Represents a simple ToDo task.
 */
public class ToDo extends Task {
```

caando commented on src/main/java/duck/Ui.java:
```diff
@@ -0,0 +1,128 @@
+package duck;
+import java.util.Scanner;
+
+/**
+ * Handles user interactions, including reading input and displaying messages.
+ */
```
I like the thorough javadoc for all classes and functions!

caando commented on src/main/java/duck/Parser.java:
```diff
@@ -0,0 +1,252 @@
+ */
+public class Parser {
+
+    /**
+     * Constructs a Parser instance.
+     */
+    public Parser() {
+    }
```
Should empty constructors be removed?


```suggestion
```

caando commented on src/main/java/duck/Event.java:
```diff
@@ -0,0 +1,62 @@
+        super(isDone, description);
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+        Random r = new Random();
+        this.from = from.equalsIgnoreCase("now") ? LocalDateTime.now()
+                : from.equalsIgnoreCase("later") ? LocalDateTime.now().plusMinutes(r.nextInt(10000))
+                : LocalDateTime.parse(from.trim(), formatter);
+        this.to = to.equalsIgnoreCase("now") ? LocalDateTime.now()
+                : from.equalsIgnoreCase("later") ? LocalDateTime.now().plusMinutes(r.nextInt(10000))
+                : LocalDateTime.parse(to.trim(), formatter);
```
Should magic numbers be `static final` variables?

```suggestion
        static final int INTERVAL_BOUND = 10000;
        this.from = from.equalsIgnoreCase("now") ? LocalDateTime.now()
                : from.equalsIgnoreCase("later") ? LocalDateTime.now().plusMinutes(r.nextInt(INTERVAL_BOUND))
                : LocalDateTime.parse(from.trim(), formatter);
        this.to = to.equalsIgnoreCase("now") ? LocalDateTime.now()
                : from.equalsIgnoreCase("later") ? LocalDateTime.now().plusMinutes(r.nextInt(INTERVAL_BOUND))
                : LocalDateTime.parse(to.trim(), formatter);
```

caando commented on src/main/java/duck/MainWindow.java:
```diff
@@ -0,0 +1,62 @@
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
+ * Controller for the main GUI.
+ */
```
Should there be a new line break between imports and javadoc?

```suggestion
import javafx.scene.layout.VBox;

/**
 * Controller for the main GUI.
 */
```

caando commented on src/main/java/duck/MainWindow.java:
```diff
@@ -0,0 +1,62 @@
+
+    private Duck duck;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/user.jpeg"));
+    private Image duckImage = new Image(this.getClass().getResourceAsStream("/images/duck.jpg"));
```
I like UI you have created!

caando commented on src/test/java/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+        Task task = new Task(false, "sleep");
+        taskList.add(task);
+        assertEquals(1, taskList.size());
+        assertEquals(task, taskList.get(0));
+    }
+    @Test
+    public void deleteTaskTest() {
```
Should methods be separated by line break?

```suggestion
        assertEquals(task, taskList.get(0));
    }

    @Test
    public void deleteTaskTest() {
```

caando commented on src/main/java/duck/Parser.java:
```diff
@@ -0,0 +1,252 @@
+     *
+     * @param ui The UI instance to display the message.
+     */
+    public String unknown(Ui ui) {
+        return ui.showCommandErrorMessage();
+    }
```
Should more descriptive method names be used? I would recommend `handleXXXCommand` 

```suggestion
    public String handleUnknownCommand(Ui ui) {
        return ui.showCommandErrorMessage();
    }
```

kennethsoh replied:
Thanks for the suggestion 👍🏻

caando reviewed (changes requested):
Great job so far with iP! The code is clear and readable with only minor nitpicks.

glennliew commented on src/main/java/duck/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     * @param description The description of the task.
+     * @param by The due date of the task in "yyyy-MM-dd HHmm" format.
+     */
+    public Deadline(Boolean isDone, String description, String by) {
```
using Boolean wrapper instead of primitive boolean
```suggestion
    public Deadline(boolean isDone, String description, String by) {
```

glennliew commented on src/main/java/duck/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+        Random r = new Random();
+        this.by = by.equalsIgnoreCase("now") ? LocalDateTime.now()
+                : by.equalsIgnoreCase("later") ? LocalDateTime.now().plusMinutes(r.nextInt(10000))
```
10000 is an arbitrary constant with no explanation, could use a named constant
```suggestion
                : by.equalsIgnoreCase("later") ? LocalDateTime.now().plusMinutes(r.nextInt(MAX_RANDOM_MINUTES))
```

glennliew commented on src/main/java/duck/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return The deadline as a LocalDateTime object.
+     */
+    public LocalDateTime deadline() {
```
getters should follow the getX() format
```suggestion
    public LocalDateTime getDeadline() {
```

glennliew commented on src/main/java/duck/Duck.java:
```diff
@@ -0,0 +1,51 @@
+     * @throws UnknownCommandException If an unrecognized command is entered.
+     * @throws IOException If an error occurs while reading or writing the storage file.
+     */
+    public static void main(String[] args) throws EmptyDetailsException, UnknownCommandException, IOException {
```
main Method Should Not Throw Exceptions Directly. If an exception is thrown, it crashes the program without proper error handling. Other than this LGTM!
```suggestion
    public static void main(String[] args) {
    try {
        new Duck().run();
    } catch (IOException e) {
        System.err.println("Error accessing storage file: " + e.getMessage());
    } catch (Exception e) {
        System.err.println("An unexpected error occurred: " + e.getMessage());
    }
}
```

glennliew reviewed (commented):
Looks good! Only some minor changes needed 👍

kennethsoh committed:
Fix event LocalDateTime features

kennethsoh committed:
Restore find function

kennethsoh committed:
Add assertion to check user input

The command extracted from the input can be empty if the user provides
malformed or blank input. This could lead to unexpected behavior in
command processing. Deadline "by" time and Event "to" time can be in
the past if the user enters a datetime that is passed the current
time. This is illogical.

An empty command should not be allowed as it may cause logic errors
or failures in subsequent operations. Deadline "by" time must be in
the future, and Event "to" time must be after "from" time.

Add an assertion to check that the extracted command is not empty.
Add an assertion to check that "by" and "to" time is logical.
This helps catch invalid inputs early during development and
debugging.

Assertions are used here because they can be enabled during
development and disabled in production. If stricter enforcement
is required, an explicit exception should be used instead.

kennethsoh committed:
Refactor code to meet CodeQuality standards

The code quality is not up to standard as per
course requirements. In some parts of the code,
magic variables were used.

Redundant code and useless comments were removed.
Magic literals are avoided to best ability.
Complicated expressions were simplified.

Improving code quality will improve readability.

kennethsoh committed:
Merge pull request #1 from kennethsoh/branch-A-Assertions

Add Assertions

kennethsoh committed:
Merge branch 'master' into branch-A-CodeQuality

kennethsoh committed:
Merge pull request #2 from kennethsoh/branch-A-CodeQuality

Refactor code to meet A-CodeQuality

kennethsoh committed:
Fix storage load bug

kennethsoh committed:
Fix OOP issues

kennethsoh committed:
Add B-Snooze and scaffolding for C-Undo

kennethsoh committed:
Rename boolean var for Code Quality

kennethsoh committed:
Fix Ui issue

kennethsoh committed:
Add Ui.png

kennethsoh committed:
Setup CI

kennethsoh committed:
Fix find method and Ui bugs

kennethsoh committed:
Improve find method

kennethsoh committed:
Remove dead code

kennethsoh committed:
Add Readme

kennethsoh committed:
Add Tests

kennethsoh committed:
Fix Test bugs

kennethsoh committed:
Update README

kennethsoh committed:
Fix Ui bugs

kennethsoh committed:
Fix Ui bugs

kennethsoh committed:
Fix StorageTest

kennethsoh committed:
Remove erroneous test

kennethsoh committed:
Rectify CI workflow issues

kennethsoh committed:
Tidy workflow file

kennethsoh committed:
Remove references to undo command

kennethsoh committed:
Remove unused import

kennethsoh committed:
Improve SLAP

kennethsoh committed:
Improve user interaction

kennethsoh committed:
Rectify minor typo

kennethsoh committed:
Fix user input error handling

kennethsoh committed:
Update workflow

kennethsoh committed:
Update project README
